src/behaviour.rs

```diff
@@ -211,12 +200,101 @@ impl<P: StoreParams> Bitswap<P> {
     }
 }
 
+impl<P: StoreParams> Bitswap<P> {
+    /// Processes an incoming bitswap request.
+    fn inject_request(&mut self, channel: BitswapChannel, request: BitswapRequest) {
+        let response = match request.ty {
+            RequestType::Have => {
+                let have = self.store.contains(&request.cid).ok().unwrap_or_default();
+                if have {
+                    RESPONSES_TOTAL.with_label_values(&["have"]).inc();
+                } else {
+                    RESPONSES_TOTAL.with_label_values(&["dont_have"]).inc();
+                }
+                tracing::trace!("have {}", have);
+                BitswapResponse::Have(have)
+            }
+            RequestType::Block => {
+                let block = self.store.get(&request.cid).ok().unwrap_or_default();
+                if let Some(data) = block {
+                    RESPONSES_TOTAL.with_label_values(&["block"]).inc();
+                    SENT_BLOCK_BYTES.inc_by(data.len() as u64);
+                    tracing::trace!("block {}", data.len());
+                    BitswapResponse::Block(data)
+                } else {
+                    RESPONSES_TOTAL.with_label_values(&["dont_have"]).inc();
+                    tracing::trace!("have false");
+                    BitswapResponse::Have(false)
+                }
+            }
+        };
+        match channel {
+            BitswapChannel::Bitswap(channel) => {
+                self.inner.send_response(channel, response).ok();
+            }
+            #[cfg(feature = "compat")]
+            BitswapChannel::Compat(peer_id) => {
+                let msg = CompatMessage::Response(request.cid, response);
+                self.compat.push_back((peer_id, msg));
+            }
+        }
+    }
+
+    /// Processes an incoming bitswap response.
+    fn inject_response(
+        &mut self,
+        id: BitswapId,
+        peer: PeerId,
+        response: BitswapResponse,
+    ) -> Option<BitswapEvent> {
+        if let Some(id) = self.requests.remove(&id) {
+            match response {
+                BitswapResponse::Have(have) => {
+                    self.query_manager
+                        .inject_response(id, Response::Have(peer, have));
+                }
+                BitswapResponse::Block(data) => {
+                    if let Some(info) = self.query_manager.query_info(id) {
+                        let len = data.len();
+                        if let Ok(block) = Block::new(info.cid, data) {
+                            RECEIVED_BLOCK_BYTES.inc_by(len as u64);
+                            if let Err(err) = self.store.insert(&block) {
```

Comment:
So we are doing the inserts one by one. That is going to be really slow if the store is in synchronous = full mode.

E.g. you get a 100 byte block of metadata, but you still pay some ~10 milliseconds penalty for a full flush.

Is there a way to do this chunked? Otherwise this means that you can not operate the block store in paranoid mode anymore.